core/src/main/java/io/questdb/log/TemplateParser.java

```diff
@@ -0,0 +1,244 @@
+                    }
+                    if (keyStart == NIL) {
+                        addEnvComponent(dollarStart, dollarStart + 1, i);
+                        lastExprEnd = i;
+                    } else {
+                        int exprLen = i - keyStart;
+                        if (exprLen == 0) {
+                            throw new LogError("Missing expression at position " + keyStart);
+                        }
+                        int formatStart = keyStart + DATE_FORMAT_KEY.length();
+                        if (Chars.startsWith(originalTxt, keyStart, formatStart, DATE_FORMAT_KEY)) {
+                            addDateComponent(formatStart, i);
+                        } else {
+                            addEnvComponent(dollarStart, dollarStart + 2, i);
+                        }
+                        lastExprEnd = i + 1;
+                    }
+                    keyStart = NIL;
+                    dollarStart = NIL;
+            }
+        }
+        if (dollarStart == NIL) {
+            if (curlyBraces != 0) {
+                throw new LogError("Mismatched '{}' at position " + lastExprEnd);
+            }
+            if (locationLen - lastExprEnd > 0) {
+                addStaticComponent(lastExprEnd, locationLen);
+            }
+        } else {
+            if (keyStart != NIL) {
+                throw new LogError("Missing '}' at position " + locationLen);
+            }
+            if (locationLen - dollarStart > 1) {
+                addEnvComponent(dollarStart, dollarStart + 1, locationLen);
+            } else {
+                throw new LogError("Unexpected '$' at position " + dollarStart);
+            }
+        }
+        return this;
+    }
+
+    public void setDateValue(long dateValue) {
+        this.dateValue.set(dateValue);
+    }
+
+    public int getKeyOffset(CharSequence key) {
+        return envStartIdxs.get(key); // relative to originalTxt
+    }
+
+    public ObjList<Component> getComponents() {
+        return txtComponents;
+    }
+
+    @Override
+    public void toSink(CharSink sink) {
+        for (int i = 0, n = txtComponents.size(); i < n; i++) {
+            sink.put(txtComponents.getQuick(i));
+        }
+    }
+
+    @Override
+    public String toString() {
+        resolveSink.clear();
+        for (int i = 0, n = txtComponents.size(); i < n; i++) {
+            resolveSink.put(txtComponents.getQuick(i));
+        }
+        return resolveSink.toString();
+    }
+
+    private enum ComponentType {
+        STATIC, ENV, DATE
+    }
+
+    public static abstract class Component implements Sinkable {
+
+
+        private final ComponentType type;
+        private final CharSequence key;
+
+        private Component(ComponentType type, CharSequence key) {
+            this.type = type;
+            this.key = key;
+        }
+
+        public boolean isEnv(CharSequence key) {
+            return type == ComponentType.ENV && this.key.equals(key);
```

Comment:
generally since the arg is `CharSequence` we should be using `Chars.equals()`. This is because `equals()` call interprets `String != CharSequence` even if contents is identical